swayfmt/src/formatter/shape.rs

```diff
@@ -103,26 +103,51 @@ impl Sub for Indent {
         Indent::new(self.block_indent - rhs.block_indent)
     }
 }
-
+/// Information about the line of code currently being evaluated.
 #[derive(Copy, Clone, Debug, PartialEq, Eq, Default)]
 pub(crate) struct CodeLine {
+    /// The current number of characters in the given code line.
+    pub(crate) width: usize,
     pub(crate) line_style: LineStyle,
     pub(crate) expr_kind: ExprKind,
+    /// Used in determining `SameLineWhere` formatting.
+    pub(crate) has_where_clause: bool,
 }
 
 impl CodeLine {
+    pub(crate) fn from(line_style: LineStyle, expr_kind: ExprKind) -> Self {
+        Self {
+            width: Default::default(),
+            line_style,
+            expr_kind,
+            has_where_clause: Default::default(),
+        }
+    }
+    pub(crate) fn reset_width(&mut self) {
+        self.width = 0;
+    }
+    pub(crate) fn update_width(&mut self, new_width: usize) {
+        self.width = new_width;
+    }
+    pub(crate) fn add_width(&mut self, extra_width: usize) {
+        self.width += extra_width;
+    }
+    pub(crate) fn sub_width(&mut self, extra_width: usize) {
+        self.width -= extra_width;
+    }
     /// Update `CodeLine::line_style` with a given LineStyle.
-    pub(crate) fn update_line_style(&mut self, line_style: LineStyle) {
-        self.line_style = line_style
+    pub(crate) fn with_line_style(&mut self, line_style: LineStyle) {
```

Comment:
Perhaps these should remain `update_line_style` and `update_expr_kind` to remain consistent with `update_width` which also just assigns the given value?

In Rust, the `with` prefix is more commonly used with builder style methods (where you return a new `Self` **with** a new value) or for functions that will call a given function **with** some argument (like `Formatter::with_shape`).